lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/storage/SharedAccessSignatureCloudBlobClientProvider.java

```diff
@@ -0,0 +1,89 @@
+ * under the License.
+ */
+package org.apache.reef.runtime.azbatch.util.storage;
+
+import com.microsoft.windowsazure.storage.StorageCredentialsSharedAccessSignature;
+import com.microsoft.windowsazure.storage.blob.CloudBlob;
+import com.microsoft.windowsazure.storage.blob.CloudBlobClient;
+import com.microsoft.windowsazure.storage.blob.SharedAccessBlobPolicy;
+import org.apache.reef.runtime.azbatch.parameters.AzureStorageAccountName;
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Cloud Blob client provider that uses Azure Storage Shared Access Signature authorization.
+ */
+public class SharedAccessSignatureCloudBlobClientProvider implements ICloudBlobClientProvider {
+
+  public static final String AZURE_STORAGE_CONTAINER_SAS_TOKEN_ENV = "AZURE_STORAGE_CONTAINER_SAS_TOKEN_ENV";
+
+  private static final String AZURE_STORAGE_ACCOUNT_URI_FORMAT = "https://%s.blob.core.windows.net";
+
+  private final String azureStorageAccountName;
+  private final String azureStorageContainerSASToken;
+
+  @Inject
+  SharedAccessSignatureCloudBlobClientProvider(
+      @Parameter(AzureStorageAccountName.class) final String azureStorageAccountName) {
+    this.azureStorageAccountName = azureStorageAccountName;
+    this.azureStorageContainerSASToken = System.getenv(AZURE_STORAGE_CONTAINER_SAS_TOKEN_ENV);
+  }
+
+  /**
+   * Returns an instance of {@link CloudBlobClient} based on available authentication mechanism.
+   * @return an instance of {@link CloudBlobClient}.
+   * @throws IOException
+   */
+  @Override
+  public CloudBlobClient getCloudBlobClient() throws IOException {
+    StorageCredentialsSharedAccessSignature signature =
+        new StorageCredentialsSharedAccessSignature(this.azureStorageContainerSASToken);
+    URI storageAccountUri;
+    try {
+      storageAccountUri = new URI(String.format(AZURE_STORAGE_ACCOUNT_URI_FORMAT, this.azureStorageAccountName));
+    } catch (URISyntaxException e) {
+      throw new IOException("Failed to generate Storage Account URI", e);
+    }
+
+    return new CloudBlobClient(storageAccountUri, signature);
+  }
+
+  /**
+   * Generates a Shared Access Key URI for the given {@link CloudBlob}.
+   * @param cloudBlob   cloud blob to create a Shared Access Key URI for.
+   * @param policy      an instance of {@link SharedAccessBlobPolicy} that specifies permissions and signature's
+   *                    validity time period.
+   * @return            a Shared Access Key URI for the given {@link CloudBlob}.
+   * @throws IOException
+   */
+  @Override
+  public URI generateSharedAccessSignature(final CloudBlob cloudBlob, final SharedAccessBlobPolicy policy)
+      throws IOException {
+    try {
+      final String uri = cloudBlob.getStorageUri().getPrimaryUri().toString();
+      return new URI(uri + "?" + this.azureStorageContainerSASToken);
```

Comment:
This will probably work fine because we know the format of the PrimaryUri, but we should always try to use UriBuilder instead of string manipulation.

Reply:
good call!